Вариант 4

In [21]:
"""imports"""

import numpy as np
import pandas as pd
import scipy.stats as sps
from statsmodels.stats.diagnostic import lilliefors

In [22]:
table = pd.read_csv("/Users/mihildrozdov/PycharmProjects/mathStatistics/song_data (1).csv")
table.head()

song_name  song_popularity  song_duration_ms  \
0  Boulevard of Broken Dreams               73            262333   
1                  In The End               66            216933   
2           Seven Nation Army               76            231733   
3                  By The Way               74            216933   
4           How You Remind Me               56            223826   

   acousticness  danceability  energy  instrumentalness  key  liveness  \
0      0.005520         0.496   0.682          0.000029    8    0.0589   
1      0.010300         0.542   0.853          0.000000    3    0.1080   
2      0.008170         0.737   0.463          0.447000    0    0.2550   
3      0.026400         0.451   0.970          0.003550    0    0.1020   
4      0.000954         0.447   0.766          0.000000   10    0.1130   

   loudness  audio_mode  speechiness    tempo  time_signature  audio_valence  
0    -4.095           1       0.0294  167.060               4          0.474  
1    -6.407           0       0.0498  105.256               4          0.370  
2    -7.828           1       0.0792  123.881               4          0.324  
3    -4.938           1       0.1070  122.444               4          0.198  
4    -5.065           1       0.0313  172.011               4          0.574

In [23]:
table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18835 entries, 0 to 18834
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   song_name         18835 non-null  object 
 1   song_popularity   18835 non-null  int64  
 2   song_duration_ms  18835 non-null  int64  
 3   acousticness      18835 non-null  float64
 4   danceability      18835 non-null  float64
 5   energy            18835 non-null  float64
 6   instrumentalness  18835 non-null  float64
 7   key               18835 non-null  int64  
 8   liveness          18835 non-null  float64
 9   loudness          18835 non-null  float64
 10  audio_mode        18835 non-null  int64  
 11  speechiness       18835 non-null  float64
 12  tempo             18835 non-null  float64
 13  time_signature    18835 non-null  int64  
 14  audio_valence     18835 non-null  float64
dtypes: float64(9), int64(5), object(1)
memory usage: 2.2+ MB


Задание 1

Пусть $X$ - случайная величина, описывающая популярность песни. Тогда нулевая гипотеза формулируется следующим образом:<br>
<br>
$H_0: X$ имеет нормальное распределение.<br>
<br>
Альтернативная гипотеза выглядит так:<br>
<br>
$H_1: X$ имеет распределение, отличное от нормального.<br>
<br>
Для проверки этого утверждения будем использовать два критерия согласия - критерий Лиллиефорса и критерий Шапиро-Уилка. Уровень значимости для обоих критериев примем равным $\alpha = 0.05$.

Статистика критерия Лиллиефорса вычисляется по формуле $D_n = \underset{x \in \mathbb{R}}{\sup} |\widehat{F_n}(x) - F_0(x)|$, где $\widehat{F_n}(x)$ - эмпирическая функция распределения выборки в точке $x$, $F_0(x)$ - теоретическая функция распределения случайной величины, согласие с законом распределения которой тестируется. В нашем случае $F_0(x)$ - это функция нормального распределения, для которого математическое ожидание и дисперсия оцениваются как выборочное среднее и выборочная несмещенная дисперсия по имеющейся выборке популярности песни соответственно. Объем имеющейся выборки обозначим как $n$.

Тест Лиллиефорса является расширением критерия согласия Колмогорова на случай неизвестных параметров тестируемого распределения. Применение этих критериев корректно в случае непрерывных функций распределения $F(x)$ и $F_0(x)$. В нашем случае мы предполагаем непрерывность величины $X$ в силу специфики данных, и непрерывность функции нормального распределения $F_0(x)$ очевидна.

Нулевая гипотеза отвергается на уровне значимости $\alpha$, если значение статистики $D_n$, посчитанное по имеющей выборке, превышает критическое значение распределения Лиллиефорса (при истинной нулевой гипотезе), соответствующее уровню значимости $\alpha$. Данные критические значения для разных объемов выборки приведены в таблицах. В нашем случае для 5% уровня значимости критическое значение определяется по формуле $0.886/\sqrt{n}$.

Для проверки гипотезы о нормальности распределения воспользуемся критерием согласия Лиллиефорса и критерием согласия Шапиро-Уилка. Уровень значимости для обоих критериев выберем равным 0.05.

Если p-value теста Лиллиефорса или Шапиро-Уилка меньше, либо равно 0.05, то отвергаем нулевую гипотезу о нормальности распределения (так как это эквивалентно тому, что значение статистики критерия по выборке превысило критическое значение).

In [24]:
0.886 / np.sqrt(18835)

0.0064558125177343805

In [25]:
testValue = round(lilliefors(x = table["song_popularity"], dist ="norm")[0], ndigits = 3)
pValue = round(lilliefors(x = table["song_popularity"], dist ="norm")[1], ndigits = 3)

print("Значение тестовой статистики и p-value теста Лиллиефорса имеют значения", testValue, pValue)

Значение тестовой статистики и p-value теста Лиллиефорса имеют значения 0.062 0.001


In [26]:
mean = (table["song_popularity"] - np.mean(table["song_popularity"]))
variable = np.sqrt(np.var(table["song_popularity"], ddof = 1))
print(sps.kstest(mean / variable, "norm")[0] == lilliefors(table["song_popularity"], dist ="norm")[0])

True


Значение статистики для критерия Колмогорова совпадает с полученным выше значением, где данные были отскалированы на выборочное среднее и несмещенную дисперсию, посчитанные по выборке.

Таким образом, значение статистики по выборке составляет 0.062, что превышает критическое значение 0.006 (аналогично, p-value 0.001 меньше, чем 0.05), что означает отвержение нулевой гипотезы на уровне значимости 0.05. Другими словами, с максимальной вероятностью ошибки первого рода в 0.05 можно утверждать, что популярность песни имеет распределение, отличное от нормального.

Тестовая статистика критерия Шапиро-Уилка имеет вид $\dfrac{(\sum_{i=1}^{n} a_i x_{(i)})^2}{\sum_{i=1}^{n} (x_i - \overline{x})^2}$, где $x_{(i)}$ - i-я порядковая статистика по выборке (i-е по порядку значение популярности песни при упорядочении выборки по возрастанию), $\overline{x} = \dfrac{\sum_{i=1}^{n} x_i}{n}$ - выборочное среднее, $a_i$ - коэффициенты, определяемые по специальным формулам и таблицам. <br>
<br>

Нулевая гипотеза отвергается на уровне значимости $\alpha$, если значение статистики $W$, посчитанное по имеющейся выборке, превышает некоторое критическое значение (посчитанное при истинной нулевой гипотезе), соответствующее уровню значимости $\alpha$.
<br>
Теперь посчитаем значение статистики и p-value теста:

In [8]:
testValue = round(sps.shapiro(table["song_popularity"])[0], ndigits = 3)
pValue = round(sps.shapiro(table["song_popularity"])[1], ndigits = 3)

print("Значение тестовой статистики и p-value теста Шапиро-Уилка имеют значения", testValue, pValue)

Значение тестовой статистики и p-value теста Шапиро-Уилка имеют значения 0.973 0.0


/Users/mihildrozdov/PycharmProjects/mathStatistics/venv/lib/python3.8/site-packages/scipy/stats/_morestats.py:1816: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


In [9]:
sps.shapiro(table["song_popularity"])

/Users/mihildrozdov/PycharmProjects/mathStatistics/venv/lib/python3.8/site-packages/scipy/stats/_morestats.py:1816: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


ShapiroResult(statistic=0.9732723832130432, pvalue=0.0)

Значение p-value для критерия Шапиро-Уилка составляет около 0, что меньше 0,05. Следовательно, с максимальной вероятностью ошибки первого рода в 0,05 мы можем отвергнуть нулевую гипотезу о нормальном распределении популярности песни.

Таким образом, мы отвергаем гипотезу о нормальности популярности песни в обоих случаях на уровне значимости 0,05. Следовательно, нельзя аппроксимировать популярность песни нормальным распределением.

Задача 2

In [10]:
borderDuration = round(np.quantile(table["song_duration_ms"], 0.6))
print(borderDuration)

221512


Для проверки значимости различий в рейтинге популярности "длинных" и "коротких" песен используется тестирование гипотезы о равенстве выборок из одной генеральной совокупности, то есть гипотеза однородности. Пусть $X$ - случайная величина, соответствующая популярности коротких песен, $Y$ - случайная величина, соответствующая популярности длинных песен. Разбиение на длинные и короткие песни производится по выборочному квантилю уровня 0.6 по имеющейся выборке продолжительности песен. Гипотезы формулируются следующим образом:<br>
<br>
$H_0: F(x) = G(x)$ для всех $x \in \mathbb R$

$H_1: F(x) \neq G(x)$ для всех $x \in \mathbb R$,<br>
<br>
где $F(x)$ и $G(x)$ - функции распределения случайных величин $X$ и $Y$ соответственно, причем эти функции распределения непрерывны. <br>
<br>
Для тестирования используется критерий однородности Колмогорова. Статистика критерия имеет вид $D_{m,n} = \overset{}{\underset{x \in \mathbb R}{sup}} | G^*_m(x, Y_{[m]}) - F^*_n(x, X_{[n]})|$, где $F^*_n$ и $G^*_m$ - эмпирические функции распределения, построенные по выборкам $X_{[n]}$ и $Y_{[m]}$ соответственно. Если гипотеза $H_0$ верна, то статистика критерия имеет асимптотическое распределение Колмогорова. Критическая область определяется квантилем Колмогорова $d_{1 - \alpha}$.<br>
<br>
Рассчитывается p-value теста. Если оно меньше или равно 0.05, то гипотеза $H_0$ отвергается на уровне значимости 0.05.

In [11]:
popularityLong = table[table["song_duration_ms"] > borderDuration]["song_popularity"]
popularityShort = table[table["song_duration_ms"] <= borderDuration]["song_popularity"]

print(len(popularityShort) + len(popularityLong) == len(table["song_popularity"]))

True


In [14]:
print(popularityLong[:5])
print(popularityShort[:5])

0    73
2    76
4    56
5    80
8    80
Name: song_popularity, dtype: int64
1    66
3    74
6    81
7    76
9    81
Name: song_popularity, dtype: int64


In [15]:
print(sps.ks_2samp(data1 = popularityShort, data2 = popularityLong)[0])

0.03172285638439076


In [16]:
testValue = round(sps.ks_2samp(data1 = popularityShort, data2 = popularityLong)[0], ndigits = 3)
pValue = round(sps.ks_2samp(data1 = popularityShort, data2 = popularityLong)[1], ndigits = 3)

print("Значение тестовой статистики и p-value критерия однородности Колмогорова имеют значения", testValue, pValue)

Значение тестовой статистики и p-value критерия однородности Колмогорова имеют значения 0.032 0.0


Таким образом, полученное p-value меньше 0.05, что говорит о том, что нулевая гипотеза отвергается на уровне значимости 0.05. Следовательно, можно с уверенностью утверждать, что распределение популярности песен зависит от их продолжительности.

Второй способ тестирования - критерий однородности Андерсона-Дарлинга.<br>
<br>
Статистика критерия имеет вид (для непрерывных функций распределения $F(x)$ и $G(x)$ в нулевой гипотезе выше):<br>
<br>
$A_2 = \dfrac{1}{mn}\sum_{i=1}^{n+m-1} \dfrac{(M_i (m+n) - mi)^2}{i(m+n-i)}$,<br>
<br>
где $M_i$ - число элементов первой выборки, меньших или равных i-му элементу вариационного ряда объединенной выборки.<br>
<br>
Нулевая гипотеза отвергается на уровне значимости 0.05 при превышении статистикой критерия соответствующего критического значения.<br>
<br>
Как и в предыдущем случае, рассчитаем p-value теста.

In [18]:
print(sps.anderson_ksamp(samples = [popularityShort, popularityLong]))

first = round(sps.anderson_ksamp(samples = [popularityShort, popularityLong])[0], ndigits = 3)
second = round(sps.anderson_ksamp(samples = [popularityShort, popularityLong])[2], ndigits = 3)
print("Значение тестовой статистики критерия однородности Андерсона-Дарлинга и p-value критерия имеют значения", first, second)

Anderson_ksampResult(statistic=7.588853157969736, critical_values=array([0.325, 1.226, 1.961, 2.718, 3.752, 4.592, 6.546]), pvalue=0.001)
Значение тестовой статистики критерия однородности Андерсона-Дарлинга и p-value критерия имеют значения 7.589 0.001


/var/folders/9f/w41bspxs7clddh3nwh3bpfcc0000gn/T/ipykernel_3011/3490888465.py:1: UserWarning: p-value floored: true value smaller than 0.001
  print(sps.anderson_ksamp(samples = [popularity_short, popularity_long]))
/var/folders/9f/w41bspxs7clddh3nwh3bpfcc0000gn/T/ipykernel_3011/3490888465.py:3: UserWarning: p-value floored: true value smaller than 0.001
  first = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[0], ndigits = 3)
/var/folders/9f/w41bspxs7clddh3nwh3bpfcc0000gn/T/ipykernel_3011/3490888465.py:4: UserWarning: p-value floored: true value smaller than 0.001
  second = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[2], ndigits = 3)


Таким образом, значение статистики критерия составило 7.589, что превышает критическое значение для 5% уровня значимости, равное 1.961 (аналогично, p-value теста составило величину меньше 0.001 < 0.05). Из этого следует, что нулевая гипотеза отвергается на уровне значимости 0.05. Можно с уверенностью утверждать, что выборки популярности длинных и коротких песен получены из разных генеральных совокупностей и имеют разные распределения с максимальной вероятностью ошибки первого рода в 0.05.

In [19]:
print(len(popularityShort), len(popularityLong))
print(np.mean(popularityShort), np.mean(popularityLong))
print(np.var(popularityShort), np.var(popularityLong))

11301 7534
53.26245465003097 52.586010087602865
483.6359224412007 473.85210584858413


В данном случае заключения двух критериев совпадают, и на уровне значимости 5% можем утверждать, что есть принципиальные отличия в популярности длинных и коротких песен.

Задача 3

Рассмотрим две случайные величины: $Z$ - популярность песни, и $D$ - ее продолжительность. Зависимость популярности песни от продолжительности означает, что при изменении значения $D$ меняется распределение $Z$, то есть условная функция распределения $F(Z | D)$ зависит от значений $D$. Введем теперь случайную величину $D_{\text{binar}} = 1, \ \text{если} \ D > 221512, \ \text{и} \ D_{\text{binar}} = 0 \ \text{иначе}$. Очевидно, это измеримая функция от случайной величины $D$. Заметим теперь, что в задаче 2 мы протестировали фактически зависимость случайных величин $Z$ и $D_{\text{binar}}$, так как проверка отличия распределения $Z$ в зависимости от того, длинная песня ($D_{\text{binar}} = 1$) или короткая ($D_{\text{binar}} = 0$), эквивалентна фактически проверке утверждения, что распределение $Z$ зависит от значений $D_{\text{binar}}$. Мы получили отвержение нулевой гипотезы в задаче 2, и заключили, что распределение $Z$ принципиально отличается в случаях $D_{\text{binar}} = 1$ и $D_{\text{binar}} = 0$. Другими словами, случайные величины $Z$ и $D_{\text{binar}}$ зависимы по определению. Так как $D_{\text{binar}}$ - это функция от величины $D$, то при изменении значений $D$ меняются значения и $D_{\text{binar}}$, а значит меняется и распределение $Z$, то есть величины $Z$ и $D$ тоже зависимы (по вероятностному определению) - если песня "короткая", то популярность будет иметь другое распределение, чем если бы песня была "длинной". Формально, тестирование зависимости здесь эквивалентно тестированию утверждений в задаче 2, то есть гипотеза о независимости эквивалентна конкретной гипотезе об однородности (хотя стоит отметить, что если бы мы не отвергли гипотезу об однородности в задаче 2, это еще не означало бы независимость. Но если мы хотя бы для одного порога отвергаем гипотезу об однородности, то это уже означает наличие зависимости величин $Z$ и $D$ по определению).